NetKet/Hilbert/bosons.hpp

```diff
@@ -157,7 +169,31 @@ class Boson : public AbstractHilbert {
     }
   }
 
-  const AbstractGraph &GetGraph() const noexcept override { return graph_; }
+  Boson &operator*=(const Boson &rhs) {
+    if (LocalSize() != rhs.LocalSize()) {
+      throw RuntimeError{"Cannot multiply the given hilbert spaces"};
+    }
+    nsites_ += rhs.Size();
+    if (constraintN_ && rhs.constraintN_) {
+      SetNbosons(nbosons_ + rhs.nbosons_);
+    } else {
+      constraintN_ = false;
```

Comment:
In the case where there is a constraint only on the left or right, maybe we could add a short print message, as it's most likely an error on the part of the user?